File: question-engine/sdk/examples/typescript/QuestionEngineClient.example.ts
export type QuestionEngineClientOptions = {
  baseUrl: string;
  headers?: Record<string, string>;
};

export type ProcessingJobInput = {
  paperFile: File | Blob;
  answerFile?: File | Blob;
  stage?: string;
  subject?: string;
  grade?: string;
  region?: string;
  year?: string;
  title?: string;
};

export class QuestionEngineClient {
  private readonly baseUrl: string;
  private readonly headers: Record<string, string>;

  constructor(options: QuestionEngineClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, "");
    this.headers = options.headers ?? {};
  }

  capabilities() {
    return this.getJson("/api/capabilities");
  }

  engine() {
    return this.getJson("/api/engine");
  }

  processingJob(jobId: string) {
    return this.getJson(`/api/capabilities/question-processing/jobs/${encodeURIComponent(jobId)}`);
  }

  questionPackage(jobId: string) {
    return this.getJson(`/api/capabilities/question-processing/jobs/${encodeURIComponent(jobId)}/question-package`);
  }

  async createProcessingJob(input: ProcessingJobInput) {
    const form = new FormData();
    form.append("paperFile", input.paperFile);
    if (input.answerFile) form.append("answerFile", input.answerFile);
    for (const key of ["stage", "subject", "grade", "region", "year", "title"] as const) {
      if (input[key]) form.append(key, input[key] ?? "");
    }
    const response = await fetch(`${this.baseUrl}/api/capabilities/question-processing/jobs`, {
      method: "POST",
      headers: this.headers,
      body: form,
    });
    return this.parse(response);
  }

  private async getJson(path: string) {
    const response = await fetch(`${this.baseUrl}${path}`, {
      headers: { ...this.headers },
    });
    return this.parse(response);
  }

  private async parse(response: Response) {
    const text = await response.text();
    const body = text ? JSON.parse(text) : null;
    if (!response.ok) {
      throw new Error(body?.detail || body?.error || text || `HTTP ${response.status}`);
    }
    return body;
  }
}

File: question-engine/sdk/examples/java/QuestionEngineClientExample.java
package com.aigeneration.questionengine.sdk;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;

public class QuestionEngineClientExample {
    private final String baseUrl;
    private final Map<String, String> headers;
    private final HttpClient client;

    public QuestionEngineClientExample(String baseUrl) {
        this(baseUrl, Map.of());
    }

    public QuestionEngineClientExample(String baseUrl, Map<String, String> headers) {
        this.baseUrl = stripTrailingSlash(baseUrl);
        this.headers = headers;
        this.client = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(5))
                .build();
    }

    public String capabilities() throws IOException, InterruptedException {
        return get("/api/capabilities");
    }

    public String engine() throws IOException, InterruptedException {
        return get("/api/engine");
    }

    public String processingJob(String jobId) throws IOException, InterruptedException {
        return get("/api/capabilities/question-processing/jobs/" + encode(jobId));
    }

    public String questionPackage(String jobId) throws IOException, InterruptedException {
        return get("/api/capabilities/question-processing/jobs/" + encode(jobId) + "/question-package");
    }

    private String get(String path) throws IOException, InterruptedException {
        HttpRequest.Builder builder = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .timeout(Duration.ofSeconds(30))
                .GET();
        headers.forEach(builder::header);
        HttpResponse<String> response = client.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("Question Engine request failed: HTTP " + response.statusCode() + " " + response.body());
        }
        return response.body();
    }

    private static String stripTrailingSlash(String value) {
        return value == null ? "" : value.replaceAll("/+$", "");
    }

    private static String encode(String value) {
        return java.net.URLEncoder.encode(value, java.nio.charset.StandardCharsets.UTF_8);
    }
}
